ops/testing.py

```diff
@@ -3270,6 +3353,38 @@ def get_notices(
         user_id: Optional[int] = None,
         types: Optional[Iterable[Union[pebble.NoticeType, str]]] = None,
         keys: Optional[Iterable[str]] = None,
-        after: Optional[datetime.datetime] = None,
     ) -> List[pebble.Notice]:
-        raise NotImplementedError(self.get_notices)
+        # Similar logic as api_notices.go:v1GetNotices in Pebble.
+
+        filter_user_id = 0  # default is to filter by request UID (root)
+        if user_id is not None:
+            filter_user_id = user_id
+        if select is not None:
+            if user_id is not None:
+                raise self._api_error(400, 'cannot use both "select" and "user_id"')
+            filter_user_id = None
+
+        if types is not None:
+            types = {(t.value if isinstance(t, pebble.NoticeType) else t) for t in types}
+        if keys is not None:
+            keys = set(keys)
+
+        notices = [notice for notice in self._notices.values() if
```

Comment:
it seems like a dict on (uid, type, key) is not a great structure for notices given how you are querying them

I suppose dicts are Ordered, and thus you default to first creation ordering, which will be close to sorted.
I dont know if it is useful to leave a note as to why a dict is "good enough", but it was defintely something I noted in the attribute type definitions, and wanted to bring up.

I guess the idea is that count will always be small (especially while testing) and when it isnt you could use a real db and do queries.

How is it implemented in Pebble? Do notices persist across restarts? Or are they only ephemeral? If you get 5000 notices, does get_notices still have to read all of them to apply a filter, or is there some filtering that is done by indexes/storage patterns?

Reply:
We discussed this in the original design: there are likely to be only a small number of unique notices, so we figured just iterating through the dumb data structure is okay for now. The dict is just for uniquification, not indexing. So this Python code mimics the Go code. If and when we want to improve that in Pebble itself, we can (though we still probably don't need it here for the test harness).